### ISSUE TITLE
Do we still need function makeMap()?

### ISSUE THREAD

**Issue by xml-project** [xproc/3.0-specification#670]
I am not sure we still need this function. We can convert a document into a map using p:cast-content-type. The only special thing seems to be, that keys are forced to be QNames.

Any opinion: Keep it or remove it?

**Comment by eriksiegel** [comment]
I'm in favor of removing. Seems a bit unnecessary. 

But does anybody remember why it was added at all?

**Comment by xml-project** [comment]
If I remember right, this function was already defined in XProc 2.0, where we did not have the current concept of JSon documents and p:cast-content-type. This made me think, that is simply a legacy function.

**Comment by gimsieke** [comment]
Since in legacy pipelines (and probably also in future native XProc 3.0 pipelines), some people construct parameter sets as XML documents, we need to keep this conversion functionality. I’m not sure whether casting a `c:param-set` to a JSON document and passing this JSON document as a `parameter` option works even if `p:cast-content-type` were specified to “do the right thing“ when casting a `c:param-set` to JSON.

`p:cast-content-type`’s conversion from XML to JSON is currently defined as implementation-defined except for the case that the input is the XPath 3.1 XML representation of JSON.

But even if we added that `p:cast-content-type` also converts a `c:param-set` into a JSON document (consisting of a name/value associative array) using the `@name` attributes as keys and the `@value` attributes as values, we’d end up with a JSON document and not with an XPath 3.1 map.

How we can use this JSON document to pass it to a `parameters` option of a step that expects such a map?

Is this supposed to work? Will the JSON document automagically be converted into a map for `parameters`?
```xml
<p:identity>
  <p:with-input port="source">
    <c:param-set>
      <c:param name="foo" value="bar"/>
    </c:param-set>
  </p:with-input>
</p:identity>
<p:cast-content-type content-type="text/json"/>
<p:xslt parameters="{.}">
  <p:with-input port="source" pipe="result@other-step"/>
  <p:with-input port="stylesheet" pipe="xsl@this-pipeline"/>
</p:xslt>
```
With `p:make-map()`, this can be written as follows:
```xml
<p:identity>
  <p:with-input port="source">
    <c:param-set>
      <c:param name="foo" value="bar"/>
    </c:param-set>
  </p:with-input>
</p:identity>
<p:xslt parameters="{p:make-map(/)}">
  <p:with-input port="source" pipe="result@other-step"/>
  <p:with-input port="stylesheet" pipe="xsl@this-pipeline"/>
</p:xslt>
```

**Comment by xml-project** [comment]
@gimsieke I see the problem with param documents, but I think you have to change them manually (or semi automatically) anyway.
What I missed is why you use `p:identity` to construct a param-set document and then convert it to a map using the function. Wouldn't it be easier to construct the map directly instead? What did I miss?

**Comment by gimsieke** [comment]
The identity step is only there to produce a `c:param-set` for demonstration purposes. 

In transpect pipelines a `c:param-set` will typically constructed by an XSLT step ([cascade/xpl/paths.xpl](https://github.com/transpect/cascade/blob/master/xpl/paths.xpl) using [cascade/xsl/paths.xsl](https://github.com/transpect/cascade/blob/master/xsl/paths.xsl) or another XSLT that imports paths.xsl). Once constructed, it can, in principle, be passed around as a map (after being written to disk for debugging purposes). However, there is a [step](https://github.com/transpect/xproc-util/tree/master/resolve-params/xpl) that resolves references to other parameters using manipulation of a `c:param-set` document. I cannot fathom which custom manipulations of a param set there are in other steps. I know that sometimes we use `c:param-set` on an ordinary document input port and we will likely continue doing so. But the number of occurrences in our pipelines is relatively small, these operations are typically well encapsulated, and there are tests. 

Since it will be possible to import user-defined XSLT functions, we might write our own `tr:parameter-map()` function if we prefer to continue manipulating `c:param-set` documents. Therefore I’m not insisting that we keep `p:make-map()`, but I maintain that there’s a use case for this function.

**Comment by xml-project** [comment]
I am not opposed to keeping it, but I saw now use-case. If I understand your argument right, you do not want to touch the stylesheet, and the stylesheet produces p:param-set. Right? This counts as a use-case and I am not opposed to keeping it. But currently this function is imho dramatically undefined in the specs. Some work have to be done either there or in the p:cast-content-type.

My suggestion would be: Remove the function and introduce the c:param-set -> map conversion as mandatory in p:cast-content-type. 

A topic for Thursday?

**Comment by gimsieke** [comment]
It’s not only that we produce them with XSLT, we also process them as normal documents (occasionally). Yes, let’s discuss it tomorrow.

And I just saw that the result of `p:cast-content-type` to JSON [is actually a map](http://spec.xproc.org/master/head/xproc/#json-documents) or an array so we would only need to specify that, in addition to the XPath 3.1 `fn:json-to-xml()` representation, also a `c:param-set` document will be cast to a map.

So this should actually work?

```xml
<p:identity>
  <p:with-input port="source">
    <c:param-set>
      <c:param name="foo" value="bar"/>
    </c:param-set>
  </p:with-input>
</p:identity>
<p:cast-content-type content-type="application/json"/><!-- returns a map -->
<p:xslt parameters=".">
  <p:with-input port="source" pipe="result@other-step"/>
  <p:with-input port="stylesheet" pipe="xsl@this-pipeline"/>
</p:xslt>
```

**Comment by xml-project** [comment]
👍

**Comment by xml-project** [comment]
Resolution on Dec. 13th call: Remove the function and make this one of the mandatory conversions in p:cast-content-type.

**Comment by xml-project** [comment]
Fixed with pr #679
